### PR TITLE
arch: arm: aarch32: USE_SWITCH support for Cortex-A/-R

### PR DESCRIPTION
**Pull request by ibirnbaum**
This PR introduces support for CONFIG_USE_SWITCH on non-Cortex-M aarch32 CPUs as a pre-condition for SMP support.

Tested on actual Zynq hardware targets and with the twister kernel testsuite on qemu_cortex_a9.
At the time being, USE_SWITCH_SUPPORTED is only set for Cortex-A CPUs as this is the type of CPU I could best test this implementation on. CPU-related ifdefs in the code always consider Cortex-R as well.

### CODE REVIEW

File: arch/arm/core/aarch32/thread.c
Comment (andyross) [inline arch/arm/core/aarch32/thread.c:143]: Nitpick: I know this naming is preexisting (I think arc did it first, with sparc and arm64 getting it via osmosis?), but it's a little confusing.  This isn't a function provided by the arch layer to implement "get_next_switch_handle", it's a **wrapper** being used internal to the architecture to save off the current thread pointer prior to the context switch in an interrupt handler.  "z_get_next_switch_handle()" is in fact a portable (not arch-layer) scheduler function that just returns the context to which the ISR should return.

I'd suggest renaming this rather than continuing the cut/paste: "arm32_next_switch_handle()" or something?

File: arch/arm/core/aarch32/cortex_a_r/exc_exit.S
Comment (andyross) [inline arch/arm/core/aarch32/cortex_a_r/exc_exit.S:106]: I'm a little sad to see this still being implemented in terms of PendSV.  Are you guys aware of the CONFIG_SWAP_NONATOMIC issue?  The problem with using a low priority exception handler is that if you flag it and then release interrupts, you may **actually** be interrupted by a higher priority interrupt, which now has to run in a context where the kernel has set things up for a new _current thread expecting a context switch that didn't happen, and thus the thread that was interrupted is not actually what it thinks is _current!

There are a handful of workarounds in the scheduler for this condition, and almost every time I change something I manage to trip over this same edge case.  Would be nice to see it go away where possible. :)

**Review by andyross — DISMISSED** [review]
A naming nitpick and a design point (a pet peeve I mention every time someone in the ARM world gets near context switching, heh), but nothing that should block merge.

File: arch/arm/core/aarch32/thread.c
Comment (ibirnbaum) [reply arch/arm/core/aarch32/thread.c:143]: I don't know where the chain starts when it comes to this particular function's naming, but I did in fact extend it from arm64 to arm. I'll change the name of this wrapper function in order to resolve that confusion.

File: arch/arm/core/aarch32/cortex_a_r/exc_exit.S
Comment (ibirnbaum) [reply arch/arm/core/aarch32/cortex_a_r/exc_exit.S:106]: I hadn't heard of this issue before, but it sure sounds plausible that the SVC context can be interrupted by an IRQ. Maybe this could be avoided by using an IPI targeting the local CPU so that the separate SVC context isn't required any more? I'll give that a try, at least as a configurable alternative to what we have right now.

Comment (andyross) [reply arch/arm/core/aarch32/cortex_a_r/exc_exit.S:106]: It's not that the SVC can itself be interrupted (IIRC arm handlers start masked).  It's that it's priority is lower than hardware interrupts.  So if a timer (or whatever) fires while you have the scheduler lock and are deciding to context switch, and you call arch_switch() having made that decision, the next code to run is the timer interrupt, not the pendsv handler.

Now that timer interrupt, for example, decides to unblock a thread or whatever, so it needs to do some work checking the current/interrupted thread.  But that's now broken, because the "_current" variable for this CPU now reflects one thread (the one the scheduler decided to context switch to) but the thread that actually got interrupted is a **different** one (the thread from which the switch was going to happen).  And this causes bugs.  If you dig around in kernel/sched.c, you'll see a places where we have to stash off the "pending_current" thread pointer and then check underneath a kconfig to see if we hit this condition.  It's sort of a mess.

I tried to get folks to at least agree to use a highest-priority trap, but apparently that's difficult because someone (on Nordic, maybe?) had some non-zephyr handlers that needed to be able to set a PendSV in interrupt context and be guaranteed that the code wouldn't run until just before return to thread mode.

Really the right solution here is to abandon the exception handler technique and use a musical-chairs context switch like other architectures, IMHO.

But none of that is a crisis for this PR: as long as you have CONFIG_SWAP_NONATOMIC set, the scheduler will handle it correctly and the only cost is extra debugging for me on ARM every time I trip over it while making changes.

**Review by bbolen — CHANGES_REQUESTED** [review]

File: arch/arm/core/aarch32/cortex_a_r/exc_exit.S
Comment (ibirnbaum) [reply arch/arm/core/aarch32/cortex_a_r/exc_exit.S:106]: I modified the arch/Kconfig once more so that SWAP_NONATOMIC is always selected if USE_SWITCH is selected. The use of USE_SWITCH is now no longer tied to the CPU, I've moved it to the board level so that it can be de-selected as both methods work on aarch32.

File: arch/arm/core/aarch32/cortex_a_r/exc_exit.S
Comment (ibirnbaum) [reply arch/arm/core/aarch32/cortex_a_r/exc_exit.S]: done.

File: arch/arm/core/aarch32/cortex_a_r/exc_exit.S
Comment (ibirnbaum) [reply arch/arm/core/aarch32/cortex_a_r/exc_exit.S]: done.

File: arch/arm/core/aarch32/swap_helper.S
Comment (ibirnbaum) [reply arch/arm/core/aarch32/swap_helper.S]: done.

**Comment by ibirnbaum** [comment]
Added another change in kernel/idle.c: when the CI ran the tests, it did not only run the kernel tests, which I had run locally, but also the samples. It was the dining philosophers sample that ran into a problem that the kernel tests don't seem to cover: scheduling with co-op priorities only. The demo eventually reached the state when the thread holding both forks wakes up via k_thread_timeout after calling k_sleep. At that time, the thread currently running is the idle thread, which actively has to invoke the re-scheduling once it is no longer _current. This is tied to an #ifdef which has to be extended for aarch32.

**Review by bbolen — APPROVED** [review]
Looks good.  I am curious why you would rather enable USE_SWITCH for each board instead of all AARCH32_CORTEX_A cpus?

**Review by carlocaione — CHANGES_REQUESTED** [review]
@andyross wouldn't this also be affected by the same issue as #41846 ?

**Comment by carlocaione** [comment]
@ibirnbaum do you want to keep this open?

**Comment by ibirnbaum** [comment]
> @ibirnbaum do you want to keep this open?

Yes, please. I'm currently working in other corners of the system, but I would eventually like to have SMP on the v7 CPUs, and if this style of context switch is required in order to achieve this, I'll get back to this PR eventually and try to figure out if and how behaviour identical to #41846 can be achieved on v7.

Speaking of other things I'm currently working on, could you have a look at #43545 if you have the time and/or appoint another reviewer? This would fix the issue that Zephyr can not currently be booted out of u-boot on the Zynq, this is the last blocker on the way before I can open a PR for an actual, non-simulated Zynq-based board.

**Comment by carlescufi** [comment]
> > @ibirnbaum do you want to keep this open?
> 
> Yes, please. I'm currently working in other corners of the system, but I would eventually like to have SMP on the v7 CPUs, and if this style of context switch is required in order to achieve this, I'll get back to this PR eventually and try to figure out if and how behaviour identical to #41846 can be achieved on v7.
> 
> Speaking of other things I'm currently working on, could you have a look at #43545 if you have the time and/or appoint another reviewer? This would fix the issue that Zephyr can not currently be booted out of u-boot on the Zynq, this is the last blocker on the way before I can open a PR for an actual, non-simulated Zynq-based board.

@ibirnbaum there's been talk about Cortex-M SMP support, as well as it switching to `USE_SWITCH` as well. Do you have a timeframe to pick this up again so that we decide if we should wait for this before we work on any additional functionality for Cortex-M?

**Comment by ibirnbaum** [comment]
> there's been talk about Cortex-M SMP support, as well as it switching to `USE_SWITCH` as well. Do you have a timeframe to pick this up again so that we decide if we should wait for this before we work on any additional functionality for Cortex-M?

By the time I had to move on to other things, it had just been pointed out that the check which thread to run next should only happen in one place instead of two, other than that, from my point of view, the code was done and the corresponding test results looked good. I had to stop before I could figure out what exactly the modification Carlo was asking for is. That would be the next step once I pick up work on this once again, I guess out of the three steps 'coming up with a solution', 'implementing it' and 'testing it', the first one is the one that's going to take the longest and is hardest to estimate as so far, I've spent zero time on this. While I'm about to start looking into this issue once again, I've got a vacation coming up, so it's somewhat hard to estimate a completion date.

**Comment by skalldri** [comment]
@carlescufi, @ibirnbaum: Hi all! I'm new to this party, but I figure this is a good enough place to get in touch with relevant people. 

I've been super interested in getting Zephyr SMP support working on Cortex-M so that I could give it a try on the RP2040. I had  some time on my hands so I hacked together a set of changes that allows Cortex-M processors to use the new `arch_switch()` APIs.

I'm not quite finished with my changes yet, but I think I'm reasonably close to something I could put up for review. Thankfully the Cortex-M changes are pretty separate from the Cortex-A/R changes, so I expect that my PR won't conflict with anything here.

I'm mostly posting this to make sure I don't step on anyone's toes. Would it be appropriate to create a new discussion topic to discuss general `arch_switch()` / SMP support for aarch32?

**Comment by carlescufi** [comment]
> I'm mostly posting this to make sure I don't step on anyone's toes. Would it be appropriate to create a new discussion topic to discuss general `arch_switch()` / SMP support for aarch32?

Could you please introduce yourself and your work in the #arm channel in [Zephyr's discord](https://chat.zephyrproject.org/)?
I think that's the best place to discuss collaboration

File: arch/Kconfig
Comment (ithinuel) [inline arch/Kconfig:37]: Is there a difference between defining:
- `select USE_SWITCH_SUPPORTED if CPU_AARCH32_CORTEX_A` here 
- `select USE_SWITCH_SUPPORTED` in the [`CPU_AARCH32_CORTEX_A`](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/arm/core/aarch32/Kconfig#L49-L58) definition?

Comment (povergoing) [reply arch/Kconfig:37]: I think no difference, but the 2nd one is better

**Pull request closed without merging**
